modules/system/base.nix

```diff
@@ -2,22 +2,43 @@
 
 {
   system.activationScripts.createRun.text = ''
-    if ! test -L /run; then
-      if ! grep -q '^run\b' /etc/synthetic.conf 2>/dev/null; then
+    if [[ ! -L /run ]]; then
+      if [[ -e /etc/synthetic.conf ]]; then
+        if ! grep -q '^run\b' /etc/synthetic.conf 2>/dev/null; then
           echo "setting up /run via /etc/synthetic.conf..."
-          echo -e "run\tprivate/var/run" | sudo tee -a /etc/synthetic.conf >/dev/null
-          sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -B &>/dev/null || true
-          sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -t &>/dev/null || true
-          if ! test -L /run; then
-            echo "warning: apfs.util failed to symlink /run"
-          fi
-      fi
-      if ! test -L /run; then
-          echo "setting up /run..."
-          sudo ln -sfn private/var/run /run
-      fi
-      if ! test -L /run; then
-        echo "warning: failed to symlink /run"
+          printf 'run\tprivate/var/run\n' | sudo tee -a /etc/synthetic.conf >/dev/null
+        fi
+
+        # for Catalina (10.15)
+        sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -B &>/dev/null || true
+        # for Big Sur (11.0)
+        sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -t &>/dev/null || true
+
+        if [[ ! -L /run ]]; then
+          printf >&2 '[1;31merror: apfs.util failed to symlink /run, aborting activation[0m\n'
+          printf >&2 'To create a symlink from /run to /var/run, please run:\n'
+          printf >&2 '\n'
+          printf >&2 "$ printf 'run\tprivate/var/run\n' | sudo tee -a /etc/synthetic.conf"
+          printf >&2 '$ sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -B # For Catalina\n'
+          printf >&2 '$ sudo /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -t # For Big Sur and later\n' >&2
+          printf >&2 '\n'
+          printf >&2 'The current contents of /etc/synthetic.conf is:\n'
+          printf >&2 '\n'
+          sudo sed 's/^/    /' /etc/synthetic.conf >&2
+          printf >&2 '\n'
+          exit 1
+        fi
+      else
```

Comment:
This condition definitely seems strange and inverted now.

Reply:
I added a comment to explain further but the old logic didn't check whether `/etc/synthetic.conf` existed and would always try to create the symlink at `/run`, but as we know Nix will be installed (and will be installed at `/nix` most likely) then `/etc/synthetic.conf` should always exist and we can only create the symlink when we know we're on a system without `/etc/synthetic.conf` (<10.15)

Reply:
The `runLink` check did contain the `/etc/synthetic.conf` existence check

https://github.com/LnL7/nix-darwin/pull/1139/commits/4bc361a6379b33fac1d3fae2d2d5989ef40fa57d#diff-98c72e612838af210de471a5278ebb18d17588b45d0c7f83bae45e3cda942ad9L28

Reply:
In the previous version we had a fallback that did the symlink when `! test -L /run`; I think this is why it worked for me on pre‐Catalina. Now it’s in the else branch and we only try and set it up when `test -L /run` is true, i.e. it doesn’t need setting up. That seems to not make any sense to me.

I’d also be fine just dropping support for old versions, of course, given the direction Nixpkgs is heading in (though we should add a `sw_vers` check in that case).

Reply:
I believe the `else` matches the `if [[ -e /etc/synthetic.conf ]]` line but I can’t confirm as I’m on my phone now

Reply:
You’re totally right, sorry!